src/middlewared/middlewared/plugins/virt/instance.py

```diff
@@ -0,0 +1,451 @@
+                        new['gputype'] = 'mdev'
+                    case 'MIG':
+                        new['gputype'] = 'mig'
+                        if not device.get('mig_uuid'):
+                            raise CallError('UUID is required for MIG')
+                        new['mig.uuid'] = device['mig_uuid']
+                    case 'SRIOV':
+                        new['gputype'] = 'sriov'
+            case _:
+                raise Exception('Invalid device type')
+        return new
+
+    async def __generate_device_name(self, device_names: List[str], device_type: str) -> str:
+        name = device_type.lower()
+        i = 0
+        while True:
+            new_name = f'{name}{i}'
+            if new_name not in device_names:
+                name = new_name
+                break
+            i += 1
+        return name
+
+    async def __validate_device(self, device, schema, verrors: ValidationErrors):
+        match device['dev_type']:
+            case 'PROXY':
+                verror = await self.middleware.call('port.validate_port', schema, device['source_port'])
+                verrors.extend(verror)
+            case 'DISK':
+                if device['source'] and device['source'].startswith('/dev/zvol/'):
+                    if device['source'] not in await self.middleware.call('virt.device.disk_choices'):
+                        verrors.add(schema, 'Invalid ZVOL choice.')
+
+    @api_method(VirtInstanceDeviceAddArgs, VirtInstanceDeviceAddResult, roles=['VIRT_INSTANCE_WRITE'])
+    async def device_add(self, id, device):
+        """
+        Add a device to an instance.
+        """
+        instance = await self.middleware.call('virt.instance.get_instance', id)
+        data = instance['raw']
+        if device['name'] is None:
+            device['name'] = await self.__generate_device_name(data['devices'].keys(), device['dev_type'])
+
+        verrors = ValidationErrors()
+        await self.__validate_device(device, 'virt_device_add', verrors)
+        verrors.check()
+
+        data['devices'][device['name']] = await self.__device_to_incus(instance['type'], device)
+        await incus_call_and_wait(f'1.0/instances/{id}', 'put', {'json': data})
+        return True
+
+    @api_method(VirtInstanceDeviceDeleteArgs, VirtInstanceDeviceDeleteResult, roles=['VIRT_INSTANCE_DELETE'])
+    async def device_delete(self, id, device):
+        """
+        Delete a device from an instance.
+        """
+        instance = await self.middleware.call('virt.instance.get_instance', id)
+        data = instance['raw']
+        if device not in data['devices']:
+            raise CallError('Device not found.', errno.ENOENT)
+        data['devices'].pop(device)
+        await incus_call_and_wait(f'1.0/instances/{id}', 'put', {'json': data})
+        return True
+
+    @api_method(VirtInstanceStateArgs, VirtInstanceStateResult, roles=['VIRT_INSTANCE_WRITE'])
+    @job()
+    async def state(self, job, id, action, force):
```

Comment:
We could go both ways. Any advantage in having separate methods with same implementation?

Reply:
I think it's much clearer that way i.e `stop/start/restart` instead of `state start/stop/restart` both in terms of understanding and consumption and also having varying parameters like what does force do in the event of `start`?